### ISSUE TITLE
Error handling: try/finally

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#501]
### Finally clause

It can be helpful to have a code block that is always executed, even if errors occur:

```xquery
let $tmp := file:create-temp-file()
return try {
  (: I/O stuff :)
} finally {
  file:delete($tmp)
}
```

For re-throwing errors, see #1912.

**Comment by michaelhkay** [comment]
I'm not sure how useful "finally" is in a declarative language with no state and no open resources that need to be closed. Your example isn't very convincing if it's only needed when handling file updates with side effects.

I'm inclined to go for a function `throw(map)`. Not `rethrow()`, because the map might be constructed for the purpose.

**Comment by ChristianGruen** [comment]
> I'm not sure how useful "finally" is in a declarative language with no state and no open resources that need to be closed. Your example isn't very convincing if it's only needed when handling file updates with side effects.

I agree that `finally` would be a concession that XQuery is also used for side-effecting use cases. If your code is deterministic, there’s no need for it.

> I'm inclined to go for a function `throw(map)`. Not `rethrow()`, because the map might be constructed for the purpose.

I also believe we should better extend the existing `fn:error` function; otherwise, it might get difficult to explain what’s the difference between these functions (most users will stick with `fn:error` anyway). I have slightly enhanced my initial comment in this issue to indicate how this could be done.

**Comment by ChristianGruen** [comment]
I suggest handling the return values of `try/catch/finally` combinations as follows:

`try` | `catch` | `finally` | Result of…
--- | --- | --- | ---
✓ | – | ✓ | If `try` & `finally` succeed, return value of `try` & `finally`
✓ | – | ↯ | If `try` succeeds and `finally` fails, return error of `finally`
↯ | ✓ | ✓ | If `try` fails and `catch` & `finally` succeed, return value of `catch` & `finally`
↯ | ✓ | ↯ | If `try` fails, `catch` succeeds and `finally` fails, return error of `finally`
↯ | ↯ | ✓ | If `try` & `catch` fail and `finally` succeeds, return value of `finally`
↯ | ↯ | ↯ | If `try`, `catch` & `finally` fail, return error of `finally`

If an expression contains multiple `catch` clauses, only the first one matters that matches.

Examples:

```xquery
(: returns 1, 3 :)
try { 1 } catch * { 2 } finally { 3 }

(: returns 2, 3 :)
try { 1 div 0 } catch * { 2 } finally { 3 }

(: returns 3 :)
try { 1 div 0 } catch * { 1 div 0 } finally { 3 }
```

**Comment by ChristianGruen** [comment]
I looked at other functional programming languages to understand how they handle results produced by `finally` blocks (if there is any available).

A language similar to ours is Microsoft’s F#, which discards final results (in this case, `1`):

```f#
try
  Some (1 / 0)
finally
  printfn "done"
  Some 1
```

I now think it is a good design choice not to pass on any final result, as the final purpose wil be debugging and clean up operations:

```xquery
try {
  file:append-text('file.txt', 1 to 10)
} catch * {
  'File could not be written: ' || $err:description
} finally {
  file:delete('file.txt'),
  message('File deleted')
}
```

I think we should be even stricter and raise an error if this block generates a non-empty sequence.

```f#
try {
  1 div 0
} catch * {
  message('done')
  1  (: ↯ :)
}
```

A remaining special case is `finally` raising an error. In all languages that I checked, the final error will overwrite previous results and errors.
